libvmaf/src/feature/third_party/funque/funque_ssim.c

```diff
@@ -93,5 +93,180 @@ int compute_ssim_funque(dwt2buffers *ref, dwt2buffers *dist, double *score, int
+    double* cov_xy = (double*) calloc(width * height, sizeof(double));
+
+    double* l_arr = (double*) calloc(width * height, sizeof(double));
+    double* cs_arr = (double*) calloc(width * height, sizeof(double));
+    double* ssim_arr = (double*) calloc(width * height, sizeof(double));
+
+    double* var_x_cum = *(score->var_x_cum);
+    double* var_y_cum = *(score->var_y_cum);
+    double* cov_xy_cum = *(score->cov_xy_cum);
+
+#if ENABLE_MINK3POOL
+    float cube_1minus_map = 0;
+#endif
+
+    int win_dim = (1 << n_levels);          // 2^L
+    int win_size = (1 << (n_levels << 1));  // 2^(2L), i.e., a win_dim X win_dim square
+
+    double mx, my, l, cs;
+    double sum = 0;
+    double l_sum = 0;
+    double cs_sum = 0;
+    int index = 0;
+    int index_cum = 0;
+    for(int i = 0; i < height; i++) {
+        for(int j = 0; j < width; j++) {
+            index = i * width + j;
+            mx = ref->bands[0][index] / win_dim;
+            my = dist->bands[0][index] / win_dim;
+
+            for(int k = 1; k < 4; k++) {
+                var_x_add[index] += ref->bands[k][index] * ref->bands[k][index];
+                var_y_add[index] += dist->bands[k][index] * dist->bands[k][index];
+                cov_xy_add[index] += ref->bands[k][index] * dist->bands[k][index];
+            }
+
+            var_x_cum[index] = var_x_cum[index_cum] + var_x_cum[index_cum + 1] +
+                               var_x_cum[index_cum + (width * 2)] +
+                               var_x_cum[index_cum + (width * 2) + 1] + var_x_add[index];
+            var_y_cum[index] = var_y_cum[index_cum] + var_y_cum[index_cum + 1] +
+                               var_y_cum[index_cum + (width * 2)] +
+                               var_y_cum[index_cum + (width * 2) + 1] + var_y_add[index];
+            cov_xy_cum[index] = cov_xy_cum[index_cum] + cov_xy_cum[index_cum + 1] +
+                                cov_xy_cum[index_cum + (width * 2)] +
+                                cov_xy_cum[index_cum + (width * 2) + 1] + cov_xy_add[index];
+
+            var_x[index] = var_x_cum[index] / win_size;
+            var_y[index] = var_y_cum[index] / win_size;
+            cov_xy[index] = cov_xy_cum[index] / win_size;
+
+            l_arr[index] = (2 * mx * my + C1) / ((mx * mx) + (my * my) + C1);
+            cs_arr[index] = (2 * cov_xy[index] + C2) / (var_x[index] + var_y[index] + C2);
+            ssim_arr[index] = l_arr[index] * cs_arr[index];
+#if ENABLE_MINK3POOL
+            cube_1minus_map += pow((1 - (l_arr[index] * cs_arr[index])), 3);
+#else
+            sum += (l_arr[index] * cs_arr[index]);
+            l_sum += l_arr[index];
+            cs_sum += cs_arr[index];
+#endif
+            index_cum += 2;
+        }
+        index_cum += (width * 2);
+    }
+
+#if ENABLE_MINK3POOL
+    double ssim_val = 1 - cbrt((double) cube_1minus_map / (width * height));
+    score->ssim_mean = ssim_clip(ssim_val, 0, 1);
+#else
```

Comment:
I think we can drop all the CoV pooling code and stick with mink pooled only as `essim` and `ms-essim` and simultaneously compute mean pooled values as `ssim` and `ms-ssim`.

Reply:
Mink3 pooling is added. Covariance based pooling can be removed after review